Don't default to placing each input file in a separate  band in build vrt algorithm

nyalldawson opened the pull request:
Creating a multi-band vrt is much less common vs creating a mosaic
style vrt, so set the default to the most common use case
and most predictable outcome.

The default behaviour has annoyed me for some time, but its only on watching @timlinux's latest screencast that I've finally decided to fix it :laughing:

nyalldawson committed:
[processing][gdal] Don't default to placing each input file in a separate
band in build vrt algorithm

Creating a multi-band vrt is much less common vs creating a mosaic
style vrt, so set the default to the most common use case
and most predictable outcome.

nirvn commented:
That old default also frustrated me, good move.

alexbruy reviewed (approved):
I'm fine with this move. But it should be clearly announced as some old scripts/models may rely on the old default value.

nyalldawson commented:
@alexbruy ok, valid point.

I'm trying to think of a clean way to "override" the default value in the gui only. What do you think about having a "guiDefaultValue" option which can be set through a parameter's metadata?

alexbruy commented:
> What do you think about having a "guiDefaultValue" option which can be set through a parameter's metadata?

Sounds good to me.

nyalldawson committed:
[processing] Add method to set an explicit override for the default
value of a parameter for the GUI only

This allows us freedom to change the default settings for an algorithm
shown when opening the algorithm in the toolbox/batch/model without
changing the underlying default value used for the parameter in the
raw API (which we can't do easily without potentially breaking 3rd
party scripts/plugins)

nyalldawson committed:
Fix crash on Qt >= 5.14

Master only

nyalldawson commented:
@alexbruy how's this look to you now? I've gone with an explicit getter/setter for the override instead of using the freeform metadata as I suspect we'll need to use this in a growing number of places in future, and it's a bit cleaner to have the explicit API for it...

nyalldawson committed:
Astyle

timlinux commented:
Regarding @alexbruy 's comment, wouldn't it be a good improvement (assuming it is not already the case) that when you save a model, it explicitly saves all the parameters, including defaults etc., so that if you ever change the inputs for an alg the related models still use the initially declared options?

nyalldawson commented:
@timlinux that's what already happens. It's only scripts/plugins which don't explicitly specify a value for the parameter which is an issue

alexbruy commented:
@nyalldawson oh, sorry. I've missed the notification. Looks good to me, +1 to merge.

nyalldawson commented:
Thanks @alexbruy !

alexbruy commented on src/core/processing/qgsprocessingparameters.h:
```diff
@@ -521,16 +521,60 @@ class CORE_EXPORT QgsProcessingParameterDefinition
+     * Sets the default \a value to use for the parameter in GUI widgets. Caller takes responsibility
+     * to ensure that \a value is a valid input for the parameter subclass.
+     *
+     * Usually the guiDefaultValueOverride() is a invalid variant, which indicates that the standard defaultValue()
```
Just realized that when we need to drop existing default (by setting it to invalid variant) this won't work. On the other hand this is probably an edge case and can be ignored.

nyalldawson replied:
Hm -- we could probably handle that through a null variant instead of invalid, but let's wait till we need that :)
